src/extensions/data_array.js

```diff
@@ -67,24 +96,101 @@ export default () => {
       });
     };
 
+    const createInstances = observationObjects => {
+      return new Promise(resolve => {
+        db().then(models => {
+          Promise.all(observationObjects.map(observationObj => {
+            const newReq = Object.create(req);
+            newReq.body = observationObj.body;
+            return models.createInstance(observations, newReq, exclude).then(
+              instance => {
+              observationObj.instanceId = instance.id;
+              return Promise.resolve();
+            }).catch(() => {
+              return Promise.resolve();
+            });
+          })).then(results => {
+            return resolve(results);
+          });
+        }).catch(() => {
+          return ERR.ApiError(res, 500,
+            ERR.ERRNO_INTERNAL_ERROR, ERR.INTERNAL_ERROR);
+        });
+      });
+    };
+
+    const createObservations = body => {
+      let observationObjects = [];
+      const resultIndex = body.components.indexOf(result);
+      const featureOfInterestIndex =
+        body.components.indexOf(featureOfInterest + '/id');
+      const phenomenonTimeIndex = body.components.indexOf(phenomenonTime);
+      const dataArrayCount = body[dataArrayIotCount];
+
+      for (let i = 0; i < dataArrayCount; i++) {
+        let observation = {};
+        observation.Datastream = {};
+        observation.Datastream[iotId] = body.Datastream[iotId];
+        observation.result = body.dataArray[i][resultIndex];
+        observation.phenomenonTime = body.dataArray[i][phenomenonTimeIndex];
+
+        // If FeatureOfInterest relation is not specified,
+        // it will be extracted from Locations (i.e.,
+        // Datastreams-->Things-->Locations).
+        if (featureOfInterestIndex > -1) {
+          observation.FeatureOfInterest = {};
+          observation.FeatureOfInterest[iotId] =
+            body.dataArray[i][featureOfInterestIndex];
+        }
+        observationObjects.push({
+          body: observation,
+          instanceId: null
+        });
+      }
+
+      return observationObjects;
+    };
+
     const body = req.body;
 
     // Validate input.
     // 1) The body of request must be an array.
     if (!Array.isArray(body)) {
       return ERR.ApiError(res, 400, ERR.ERRNO_BAD_REQUEST,
-                   ERR.BAD_REQUEST, 'Input must be an array');
+                          ERR.BAD_REQUEST, 'Input must be an array');
     }
 
+    let observationObjects = [];
     for (let i = 0; i < body.length; i++) {
       try {
         validateBody(body[i]);
-      } catch (ex) {
+      } catch(ex) {
         return ERR.ApiError(res, 400, ex.errno, ex.name, ex.errors);
       }
+
+      observationObjects =
+        observationObjects.concat(createObservations(body[i]));
     }
 
-    ERR.ApiError(res, 501, ERR.ERRNO_NOT_IMPLEMENTED, ERR.NOT_IMPLEMENTED);
+    const createResponse = () => {
```

Comment:
I'm wondering if that's a proper place for this function since it is specific to one endpoint. The comment in ```utils.js``` is "Module with utility shared methods"